src/nubank/gen_cert.rs

```diff
@@ -0,0 +1,216 @@
+        Some(header_value) => header_value,
+        None => return Err(GenCertError::HeaderKeyNotFound(headers.to_owned())),
+    };
+    let parsed_header_value = parse_header_value(header_value)?;
+    let (encrypted_code, sent_to) = match (
+        parsed_header_value.get("device-authorization_encrypted-code"),
+        parsed_header_value.get("sent-to"),
+    ) {
+        (Some(v1), Some(v2)) => (
+            v1.to_string().replace("\"", ""),
+            v2.to_string().replace("\"", ""),
+        ),
+        _ => {
+            return Err(GenCertError::HeaderValueKeyNotFound(
+                header_value.to_owned(),
+            ))
+        }
+    };
+
+    Ok(CodeRequestOutput {
+        url,
+        sent_to,
+        encrypted_code,
+        payload,
+        key1,
+    })
+}
+
+pub async fn exchange_certs(
+    cert_folder: &str,
+    code_request_output: CodeRequestOutput,
+    code: &str,
+) -> Result<String, GenCertError> {
+    let new_payload = build_new_payload(
+        code_request_output.payload.clone(),
+        &code_request_output.encrypted_code.clone(),
+        code,
+    );
+
+    let url = code_request_output.url.clone();
+
+    let client = reqwest::Client::new();
+    let response_str = client
+        .post(url)
+        .json(&new_payload)
+        .send()
+        .await
+        .map_err(GenCertError::ExchangeCertRequestFailed)?
+        .text()
+        .await
+        .map_err(GenCertError::ExchangeCertRequestDecodingFailed)?;
+
+    let key = code_request_output.key1.clone();
+    let cert_str = serde_json::from_str::<ExchangeCertDTO>(&*response_str)
+        .map_err(GenCertError::ExchangeCertJsonConversionFailed)?
+        .certificate;
+
+    let cert_bin = get_cert_bin(&cert_str, &key)?;
+
+    let full_path = format!("{}/{}", cert_folder, "cert.p12");
+    let _ = fs::write(&full_path, cert_bin).map_err(GenCertError::FailedToWriteCert)?;
+    Ok(full_path)
+}
+
+fn gen_private_key() -> Result<PKey<Private>, GenCertError> {
+    let rsa = Rsa::generate(2048).map_err(GenCertError::FailedToGeneratePrivateKey)?;
+    let private_key = PKey::from_rsa(rsa).map_err(GenCertError::FailedToGeneratePrivateKey);
+    private_key
+}
+
+fn build_payload(
+    cpf: &str,
+    password: &str,
+    key1: &PKey<Private>,
+    key2: &PKey<Private>,
+    device_id: String,
+) -> Result<HashMap<String, String>, GenCertError> {
+    let pub_key1 = get_public_key(key1)?;
+    let pub_key2 = get_public_key(key2)?;
+    let model = format!("PyNubank Client ({})", device_id);
+    let mut payload: HashMap<String, String> = HashMap::new();
+    payload.insert("login".to_string(), cpf.to_string());
+    payload.insert("password".to_string(), password.to_string());
+    payload.insert("public_key".to_string(), pub_key1);
+    payload.insert("public_key_crypto".to_string(), pub_key2);
+    payload.insert("model".to_string(), model);
+    payload.insert("device_id".to_string(), device_id);
+    Ok(payload)
```

Comment:
I think that it would be much more readable if we use a struct instead of a HashMap

Reply:
I agree.